### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by RoniDataloopAi**
Adding DeepLab v3 model. The branch is after publishing and testing in RC ( scope project).

### CODE REVIEW

File: adapters/deeplabv3/model_adapter.py [hunk 1]
```diff
@@ -0,0 +1,480 @@
...
+        self.model = torch.hub.load('pytorch/vision:v0.10.0', model_name, pretrained=True)
+        if weights is not None:
+            weights_filename = os.path.join(local_path, self.configuration.get('weights_filename'))
+            if os.path.isfile(weights_filename):
+                logger.info("Loading a model from {}".format(local_path))
+                num_classes = len(self.model_entity.id_to_label_map.items())
+                self.model.classifier[-1] = torch.nn.Conv2d(256, num_classes, kernel_size=(1, 1), stride=(1, 1))
+                self.model.load_state_dict(torch.load(weights_filename, map_location=self.device))
+                logger.info("Loaded custom weights {}".format(weights_filename))
+            else:
```
Comment (datafranch) [inline adapters/deeplabv3/model_adapter.py:47]: If weights are not found, raise error. If they don't configure any weights filename, then it's ok to use default.

File: adapters/deeplabv3/model_adapter.py [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
+        optimizer = optim.SGD(self.model.parameters(), lr=0.001, momentum=0.9)
+        # Decay LR by a factor of 0.1 every 7 epochs
+        exp_lr_scheduler = optim.lr_scheduler.StepLR(optimizer, step_size=3, gamma=0.1)
+
+        since = time.time()
+
+        # early stopping params
+        best_loss = np.inf
+        best_acc = 0.0
+        not_improving_epochs = 0
```
Comment (datafranch) [inline adapters/deeplabv3/model_adapter.py:178]: Shouldn't these parameters also be available for the user in config?

File: adapters/deeplabv3/model_adapter.py [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+
+                running_loss = 0.0
+                running_corrects = 0
+                total = 0
+
+                # Iterate over data
+                with tqdm.tqdm(dataloaders[phase], unit="batch") as tepoch:
+                    for batch in tepoch:
+                        inputs = torch.stack(tuple(batch[0]), 0).to(self.device)
+                        # labels = torch.stack([item['labels'] for item in batch[1]]).to(self.device)
```
Comment (datafranch) [inline adapters/deeplabv3/model_adapter.py:210]: Better to remove it if it's going to be commented.

File: adapters/deeplabv3/model_adapter.py [hunk 4]
```diff
@@ -0,0 +1,480 @@
...
+                    exp_lr_scheduler.step()
+
+                logger.info(
+                    f'Epoch {epoch}/{num_epochs} - {phase} - Loss: {epoch_loss:.4f}, Acc: {epoch_acc:.4f}, '
+                    f'Duration {(time.time() - epoch_time):.2f}')
+                plots = [dl.PlotSample(figure='loss',
+                                       legend=phase,
+                                       x=epoch,
+                                       y=epoch_loss),
+                         dl.PlotSample(figure='accuracy',
```
Comment (datafranch) [inline adapters/deeplabv3/model_adapter.py:246]: Doesn't the trainer provide any other figures? Accuracy doesn't seem to be a good metric for the task.

File: adapters/deeplabv3/model_adapter.py [hunk 5]
```diff
@@ -0,0 +1,480 @@
...
+            # self.plot_contours(img_tensor.squeeze(0).permute(1, 2, 0).cpu().numpy(), contours[0])
+
+            for polygon in contours:
+                collection.add(annotation_definition=dl.Polygon(geo=polygon, label=class_label),
+                               model_info={'name': self.model_entity.name,
+                                           'confidence': confidence,
+                                           'model_id': self.model_entity.id})
+        return collection
+
+    def convert_from_dtlpy(self, data_path, **kwargs):
```
Comment (datafranch) [inline adapters/deeplabv3/model_adapter.py:347]: If it's going to be empty, I don't think we must leave it here

File: adapters/deeplabv3/model_adapter.py [hunk 6]
```diff
@@ -0,0 +1,480 @@
...
+                    # Add to the overall true and predicted labels
+                    y_true_total.extend(y_true)
+                    y_pred_total.extend(y_pred)
+
+                data = confusion_matrix(y_true, y_pred, labels=range(len(labels)), normalize='true')
+                color_map = colors(data)
+                href_map = [[self.model_entity.dataset.platform_url for _ in range(data.shape[0])] for _ in
+                            range(data.shape[0])]
+
+                confusion = ConfusionMatrix(title="Confusion",
```
Comment (datafranch) [inline adapters/deeplabv3/model_adapter.py:393]: Ok, I think we have better metrics here. Now it makes more sense.

**Review by datafranch — CHANGES_REQUESTED** [review]

File: adapters/deeplabv3/model_adapter.py (hunk 2 above)
Comment (RoniDataloopAi) [reply adapters/deeplabv3/model_adapter.py:178]: there is a patience_epochs parameter in the config

File: adapters/deeplabv3/model_adapter.py (hunk 5 above)
Comment (RoniDataloopAi) [reply adapters/deeplabv3/model_adapter.py:347]: it raise if the function is not implemented

File: adapters/deeplabv3/model_adapter.py (hunk 1 above)
Comment (RoniDataloopAi) [reply adapters/deeplabv3/model_adapter.py:47]: i'll add it, thanks!

**Review by datafranch — APPROVED** [review]

File: adapters/deeplabv3/model_adapter.py [hunk 7]
```diff
@@ -0,0 +1,479 @@
...
+            This function is called by load_from_model (download to local and then loads)
+
+        :param local_path: `str` directory path in local FileSystem
+        """
+        self.device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+        weights = self.configuration.get('weights_filename', None)
+        model_name = self.configuration.get('model_name', 'deeplabv3_resnet50')
+        self.model = torch.hub.load('pytorch/vision:v0.10.0', model_name, pretrained=True)
+        if weights is not None:
+            weights_filename = os.path.join(local_path, self.configuration.get('weights_filename'))
```
Comment (ShadiMah) [inline adapters/deeplabv3/model_adapter.py:40]: use `weights` variables declared above

File: adapters/deeplabv3/model_adapter.py [hunk 8]
```diff
@@ -0,0 +1,479 @@
...
+            if os.path.isfile(weights_filename):
+                logger.info("Loading a model from {}".format(local_path))
```
Comment (ShadiMah) [inline adapters/deeplabv3/model_adapter.py:42]: fix logger to weights path

File: adapters/deeplabv3/model_adapter.py [hunk 9]
```diff
@@ -0,0 +1,479 @@
...
+            # Create a new id_to_label_map with 0 as 'background'
+            new_id_to_label_map = {0: 'background'}
+
+            # Adjust all other IDs by incrementing them by 1
+            for idx, label in enumerate(self.model_entity.id_to_label_map.values(), start=1):
+                new_id_to_label_map[idx] = label
+
+            # Update the model entity's id_to_label_map
+            self.model_entity.id_to_label_map = new_id_to_label_map
+            self.model_entity.label_to_id_map = new_id_to_label_map
```
Comment (ShadiMah) [inline adapters/deeplabv3/model_adapter.py:64]: Add logger that you are changing the id_to_label_map

**Review by ShadiMah — CHANGES_REQUESTED** [review]
